libtransmission/bandwidth.cc

```diff
@@ -335,3 +341,165 @@ void Bandwidth::notifyBandwidthConsumed(tr_direction dir, size_t byte_count, boo
         this->parent_->notifyBandwidthConsumed(dir, byte_count, is_piece_data, now);
     }
 }
+
+/***
+****
+***/
+
+void tr_bandwidthGroupGetLimits(
+    tr_bandwidth_group* group,
+    bool* up_limited,
+    unsigned int* up_limit,
+    bool* down_limited,
+    unsigned int* down_limit)
+{
+    *up_limit = tr_toSpeedKBps(group->bandwidth->getDesiredSpeedBytesPerSecond(TR_UP));
+    *down_limit = tr_toSpeedKBps(group->bandwidth->getDesiredSpeedBytesPerSecond(TR_DOWN));
+    *up_limited = group->bandwidth->isLimited(TR_UP);
+    *down_limited = group->bandwidth->isLimited(TR_DOWN);
+}
+
+void tr_bandwidthGroupSetLimits(
+    tr_bandwidth_group* group,
+    bool up_limited,
+    unsigned int up_limit,
+    bool down_limited,
+    unsigned int down_limit)
+{
+    group->bandwidth->setDesiredSpeedBytesPerSecond(TR_UP, tr_toSpeedBytes(up_limit));
+    group->bandwidth->setDesiredSpeedBytesPerSecond(TR_DOWN, tr_toSpeedBytes(down_limit));
+    group->bandwidth->setLimited(TR_UP, up_limited);
+    group->bandwidth->setLimited(TR_DOWN, down_limited);
+}
+
+/* Should only be called if the group does not exist */
+tr_bandwidth_group* tr_bandwidthGroupNew(tr_session* session, std::string name)
+{
+    tr_bandwidth_group* newGroup = tr_new0(tr_bandwidth_group, 1);
+    newGroup->next = nullptr;
+    newGroup->name = name;
+    newGroup->bandwidth = new Bandwidth(session->bandwidth);
+    return newGroup;
+}
+
+/* Find a bandwidth group by name. Create if it does not exist */
+tr_bandwidth_group* tr_bandwidthGroupFind(tr_session* session, std::string_view name)
+{
+    if (name.empty())
+    {
+        return nullptr;
+    }
+
+    tr_bandwidth_group* found = session->groups;
+    tr_bandwidth_group* last = nullptr;
+    while (found != nullptr && name != found->name)
+    {
+        last = found;
+        found = found->next;
+    }
+
+    if (found == nullptr)
+    {
+        found = tr_bandwidthGroupNew(session, std::string{ name });
+        if (last != nullptr)
+        {
+            last->next = found;
+        }
+        else
+        {
+            session->groups = found;
+        }
+    }
+    return found;
+}
+
+void tr_bandwidthGroupRead(tr_session* session, char const* configDir)
+{
+    tr_variant group_list;
+    char* filename = tr_buildPath(configDir, "bandwidthGroups", nullptr);
+    if (tr_variantFromFile(&group_list, TR_VARIANT_PARSE_JSON, filename, nullptr) && tr_variantIsList(&group_list))
+    {
```

Comment:
_nit_ could avoid excessive nesting by inverting the logic and returning here if !variantFromFile() || !variantIsList()